testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportAfterMethodNotAlwaysRun.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.testng.services;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import org.testng.IClassListener;
+import org.testng.ITestClass;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.AfterGroups;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.AfterSuite;
+import org.testng.annotations.AfterTest;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import static com.google.common.base.Throwables.getStackTraceAsString;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.testng.services.Listeners.reportListenerFailure;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Predicate.not;
+import static java.util.stream.Collectors.joining;
+
+public class ReportAfterMethodNotAlwaysRun
+        implements IClassListener
+{
+    private static final Set<AnnotationPredicate<?>> AFTER_ANNOTATIONS = ImmutableSet.of(
+            new AnnotationPredicate<>(AfterTest.class, not(AfterTest::alwaysRun)),
+            new AnnotationPredicate<>(AfterMethod.class, not(AfterMethod::alwaysRun)),
+            new AnnotationPredicate<>(AfterClass.class, not(AfterClass::alwaysRun)),
+            new AnnotationPredicate<>(AfterSuite.class, not(AfterSuite::alwaysRun)),
+            new AnnotationPredicate<>(AfterGroups.class, not(AfterGroups::alwaysRun)));
```

Comment:
the predicate doesn't look for "after annotations". it looks for "after annotations that lack certain attribute"

AFTER_ANNOTATIONS -> VIOLATIONS?

Reply:
Good idea, changed